Feature/add category order hint

PhilippSpo opened the pull request.

PhilippSpo commented:
@sebbulon @hajoeichler @emmenko 
import:
- added mapping for the `categoryOrderHint` column, so the csv can get validated
- added validation for the `categoryOrderHint` column

export:
- added mapping for the `categoryOrderHint` property so it gets added to the csv

mmoelli commented:
@PhilippSpo Please also add some info in the ReadMe to explain how to use the new column.

PhilippSpo commented:
@mmoelli added docs for the new column. Are they clear enough?

mmoelli commented:
looks great

sebbulon commented:
@PhilippSpo question: Does it only work with the internal category id? Or can I also use the category name or category slug to identify my category in this field?

PhilippSpo commented:
@sebbulon currently only internal category id and category name. But I already talked to @mmoelli to also use external category ids here.

hajoeichler commented on src/coffee/exportmapping.coffee:
```diff
@@ -99,6 +99,13 @@ class ExportMapping
+      categoryOrderHints = []
+      _.each categoryIds, (categoryId) ->
+        categoryOrderHints.push("#{categoryId}:#{product.categoryOrderHints[categoryId]}")
+      row[@header.toIndex CONS.HEADER_CATEGORY_ORDER_HINTS] = categoryOrderHints.join ';'
```
There are constants for the delimiters.

hajoeichler commented on src/coffee/exportmapping.coffee:
```diff
@@ -99,6 +99,13 @@ class ExportMapping
+    if @header.has(CONS.HEADER_CATEGORY_ORDER_HINTS)
+      categoryIds = Object.keys product.categoryOrderHints
+      categoryOrderHints = []
+      _.each categoryIds, (categoryId) ->
```
Can't we use `_.map` here?

hajoeichler commented on src/coffee/mapping.coffee:
```diff
@@ -112,6 +113,51 @@ class Mapping
+      [rawCatId, rawOrderHint] = rawCatOrderHint.split ':'
+      orderHint = parseFloat(rawOrderHint)
+      # check if the product is actually assigned to the category
+      if _.has(@categories.id2fqName, rawCatId)
```
I prefer the following style
```
catId = if ...
  rawcatId
else
 @categories...
```
```

hajoeichler commented on src/coffee/mapping.coffee:
```diff
@@ -112,6 +113,51 @@ class Mapping
+      orderHint = parseFloat(rawOrderHint)
+      # check if the product is actually assigned to the category
+      if _.has(@categories.id2fqName, rawCatId)
+        catId = rawCatId
```
I guess this case is covered by the last `else` already

hajoeichler commented on src/coffee/mapping.coffee:
```diff
@@ -112,6 +113,51 @@ class Mapping
+    return catOrderHints unless @hasValidValueForHeader(rawMaster, CONS.HEADER_CATEGORY_ORDER_HINTS)
+    # parse the value to get a list of all catOrderHints
+    rawCatOrderHints = rawMaster[@header.toIndex CONS.HEADER_CATEGORY_ORDER_HINTS].split GLOBALS.DELIM_MULTI_VALUE
+    for rawCatOrderHint in rawCatOrderHints
```
Please use the `underscore` library.

PhilippSpo replied:
I honestly just copy pasted this from the method above... Should I go for consistency or quality of new code?

hajoeichler commented on src/coffee/mapping.coffee:
```diff
@@ -112,6 +113,51 @@ class Mapping
+        else
+          @errors.push msg
+
+      catOrderHints[catId] = rawOrderHint
```
Even if the `orderHint` is `NAN`, you still store it here.

PhilippSpo commented:
@hajoeichler @emmenko @mmoelli @sebbulon I updated the version of the sdk and added more tests for this, also for using an external category id, category name and category slug. Are all requirements covered by the tests now?

hajoeichler commented:
:+1: and thx for the test

mmoelli commented:
looks good. 2 minor things:
* I haven't found a test for the export
* I'm missing the full explanation what kind of import options are supported

PhilippSpo commented:
@mmoelli added test for exporting and added docs for import options

mmoelli commented:
Looks great. :+1:

sebbulon commented:
:+1:
